WIP Implement Debugger Collab Functionality

Anthony-Eid opened the pull request:
This PR sets the ground work for the debugger collab functionality. There's still a lot to do, but this is being merged in because it's a good foundation for future work and this PR has gotten very large in size.

Release Notes:

- SetDebugPanelItem proto message implemented 
- UpdateDebugAdapter proto message implemented
- ProtoConversion trait implementations for dap_types
- proto definitions for dap_types


Preview:
Please note that stack frame is now rendering correctly on downstream clients 

<img width="1512" alt="Screenshot 2024-12-08 at 7 53 22 PM" src="https://github.com/user-attachments/assets/763cbb8e-0b35-4ca4-a1e4-bc3fb6c3610c" />

Anthony-Eid committed:
Initial WIP for impl FollowableItem for DebugPanelItem

Anthony-Eid committed:
Implment DebuggerThreadState proto functions

RemcoSmitsDev commented on crates/dap/src/client.rs:
```diff
@@ -31,6 +32,26 @@ pub enum ThreadStatus {
     Ended,
 }
 
+impl ThreadStatus {
```
I think we can move the **ThreadStatus** to the **debugger_ui** crate. I think I forgot to move this, when I moved a lot of code to the right spots.

Anthony-Eid committed:
Add Debug panel item variable list definition to zed.proto

Anthony-Eid committed:
Add more debug panel item messages to zed.proto

Anthony-Eid committed:
Merge branch 'debugger' into debugger-collab

Anthony-Eid committed:
WIP

Anthony-Eid committed:
Fix compile errors

Co-authored-by: Remco Smits <djsmits12@gmail.com>

Anthony-Eid committed:
More WIP

Co-authored-by: Remco Smits <djsmits12@gmail.com>

Anthony-Eid committed:
Further WIP lol

Anthony-Eid committed:
Start working on fake adapter WIP

Co-authored-by: Remco Smits <djsmits12@gmail.com>
Co-authored-by: Mikayla Maki  <mikayla@zed.dev>

RemcoSmitsDev committed:
Move process to transport struct itself

RemcoSmitsDev committed:
Add test to make sure we can send request and receive a response

RemcoSmitsDev committed:
Align other handle methods

RemcoSmitsDev committed:
Fix issues inside cargo.toml require test features inside the correct *-dependencies

RemcoSmitsDev committed:
Remove comments that are not needed

RemcoSmitsDev committed:
Add as_fake instead of downcasting

RemcoSmitsDev committed:
Clean up

RemcoSmitsDev committed:
Override get_binary method so we don't fail on install

Anthony-Eid committed:
Fix false positive clippy error

This error was a match variant not being covered when the variant wasn't possible dued
to a feature flag. I'm pretty sure this is a bug in clippy/rust-analyzer and will
open an issue on their repos

Anthony-Eid committed:
Merge branch 'debugger' into debugger-collab

Anthony-Eid committed:
Merge branch 'refactor-transport-code' into debugger-collab

Anthony-Eid committed:
Merge branch 'fake-dap' into debugger-collab

Anthony-Eid committed:
Add todo to change in dap adapter downloads

Anthony-Eid committed:
WIP Get variable to send during variable list

Anthony-Eid committed:
Get variable list from/to_proto working

Note: For some reason variable entries aren't rendering even though
everything is being sent

Anthony-Eid committed:
Fix warning messages

Anthony-Eid committed:
Fix typo

Anthony-Eid committed:
Impl stack from list from/to_proto for debug panel item

Anthony-Eid committed:
Change order of set_from_proto for debug panel item

Anthony-Eid committed:
Merge with debugger

Anthony-Eid committed:
Impl Variable list variables to/from proto funcs

Anthony-Eid committed:
Start work on Set Debugger Panel Item event

Anthony-Eid committed:
WIP with remco

Co-authored-by: Remco Smits <djsmits12@gmail.com>

Anthony-Eid committed:
Get SetDebugPanelItem proto message sending and handled

Co-authored-by: Remco Smits <djsmits12@gmail.com>

Anthony-Eid committed:
Setup UpdateDebugAdapter collab message & send live stack frame list updates

RemcoSmitsDev committed:
Use proto enums instead of hardcoded integers

RemcoSmitsDev committed:
Use read instead of update

RemcoSmitsDev committed:
Send variable list update message each time we build the entries

RemcoSmitsDev committed:
Send stack frame message when we selected the active stackframe

RemcoSmitsDev committed:
Add more mappings

RemcoSmitsDev committed:
Remove debug and rename method to be more inline with others

RemcoSmitsDev committed:
Use the correct entries to reset

RemcoSmitsDev committed:
Add tests to validate we can go and from proto ScopeVariableIndex

RemcoSmitsDev committed:
Rename test

Anthony-Eid committed:
Create UpdateAdapter ModuleList variant WIP

Anthony-Eid committed:
Change proto conversion trait to have some types return Result enums

Anthony-Eid committed:
Get clippy to pass

I removed some proto message we used in DebugPanelItem FollowableItem implmentation
because they were causing clippy errors and will need to be change in the near
future anyway

Anthony-Eid committed:
Merge branch 'debugger' into debugger-collab

Anthony-Eid committed:
Merge branch 'debugger' into debugger-collab
